test/gtest/codegen_group_conv.hpp

```diff
@@ -0,0 +1,567 @@
+            };
+            static_assert(CONV_DIR == Direction::BackwardWeights);
+            // in and out are populated with different values.
+            SetupWrw(gen_value_wrw_in, gen_value_wrw_out);
+        }
+
+        auto& handle = get_handle();
+        in_dev       = handle.Write(input.data);
+        wei_dev      = handle.Write(weights.data);
+        out_dev      = handle.Write(output.data);
+    }
+
+    void TearDown() override
+    {
+        if(test_skipped)
+            return;
+
+        auto& handle = get_handle();
+
+        if constexpr(CONV_DIR == Direction::Forward)
+        {
+            ref = ref_conv_fwd(input, weights, output, conv_desc, alpha, beta);
+            handle.ReadToVec(out_dev, output.data);
+            verify(output);
+        }
+        else if constexpr(CONV_DIR == Direction::BackwardData)
+        {
+            ref = ref_conv_bwd(input, weights, output, conv_desc, alpha, beta);
+            handle.ReadToVec(in_dev, input.data);
+            verify(input);
+        }
+        else
+        {
+            static_assert(CONV_DIR == Direction::BackwardWeights);
+            ref = ref_conv_wrw(input, weights, output, conv_desc, alpha, beta);
+            handle.ReadToVec(wei_dev, weights.data);
+            verify(weights);
+        }
+    }
+
+    GroupConvTestConfig<NDIM> conv_config;
+    miopen::ConvolutionDescriptor conv_desc;
+    tensor<T> input;
+    tensor<T> weights;
+    tensor<T> output;
+    tensor<T> ref;
+    miopen::Allocator::ManageDataPtr in_dev;
+    miopen::Allocator::ManageDataPtr wei_dev;
+    miopen::Allocator::ManageDataPtr out_dev;
+    bool test_skipped                  = false;
+    miopenTensorLayout_t tensor_layout = miopenTensorNHWC;
+    Workspace wspace{};
+
+    miopen::Scalar alpha{1.0};
+    miopen::Scalar beta{0.0};
+};
+
+template <unsigned NDIM>
+std::vector<miopenTensorLayout_t> GetLayoutValues()
+{
+    static_assert(NDIM == 2u || NDIM == 3u);
+    if constexpr(NDIM == 2u)
+    {
+        return {miopenTensorNHWC, miopenTensorNCHW};
+    }
+    else
+    {
+        return {miopenTensorNDHWC, miopenTensorNCDHW};
+    }
+}
+
+} // namespace codegen_group_conv
+
+// Test case based on 2d vs 3d dimension
+// 2d conv only support alpha 1.0
+template <unsigned ND>
+std::vector<float> GetAlphaValues()
+{
+    if constexpr(ND == 3)
+    {
+        return {1.0f, 2.2f}; /* alpha, can't be zero*/
+    }
+    else
+    {
+        return {1.0f};
+    }
+}
+
+// Test case based on 2d vs 3d dimension
+// 2d conv only support beta 0.0
+template <unsigned ND>
+std::vector<float> GetBetaValues()
+{
+    if constexpr(ND == 3)
+    {
+        return {0.0f, 3.3f};
+    }
+    else
+    {
+        return {0.0f};
+    }
+}
+
+#define DEFINE_GROUP_CONV_TEST(ndim, type, naming_type, dir, cg)                            \
+    struct GPU_GroupConv##ndim##D_##dir##_##naming_type##_##cg                              \
```

Comment:
`type` (`FP16`) must be the last in the test name. Please take a look at https://github.com/ROCm/MIOpen/wiki/GTest-development

Reply:
Also you may use kind of `lightweight` version of a new type declaration:
```C++
using GPU_GroupConv##ndim##D_##dir##_##naming_type##_##cg = GroupConvTestFix<ndim, type, Direction::dir, cg>;
```
You don't need extra brackets)